src/nodestore_attribute_helper.c

```diff
@@ -0,0 +1,209 @@
+        SLIST_FOREACH(last, &head->displayName, pointers){
+            if (!last->pointers.sle_next)
+                break;
+        };
+
+        if (last)
+            SLIST_INSERT_AFTER(last, lt, pointers);
+        else
+            SLIST_INSERT_HEAD(&head->displayName, lt, pointers);
+    }
+
+    return UA_STATUSCODE_GOOD;
+}
+
+UA_LocalizedText UA_NodeStore_getDescriptionForHead(const UA_NodeHead *head, UA_Server *server,
+                                                    const UA_NodeId *sessionId)
+{
+    UA_LocalizedText result;
+    UA_LocalizedText_init(&result);
+    LocalizedTextEntry *lt = NULL;
+
+    if (server && sessionId) {
+        size_t localeIdsSize = 0;
+        const UA_LocaleId * locales = UA_Server_getLocaleIdsForSession(server, sessionId,
+                                                                       &localeIdsSize);
+
+        /* Exact match */
+        for (size_t i = 0; i < localeIdsSize; ++i) {
+            SLIST_FOREACH(lt, &head->description, pointers) {
+                if (UA_String_equal(&locales[i], &lt->localizedText.locale)) {
+                    UA_LocalizedText_copy(&lt->localizedText, &result);
+                    return result;
+                }
+            }
+        }
+
+        /* Partial match, e.g. de-DE instead of de-CH */
+        for (size_t i = 0; i < localeIdsSize; ++i) {
+            if (locales[i].length < 2 || (locales[i].length > 2 && locales[i].data[2] != '-'))
+                continue;
+
+            UA_String requestedPrefix;
+            requestedPrefix.data = locales[i].data;
+            requestedPrefix.length = 2;
+
+            SLIST_FOREACH(lt, &head->description, pointers) {
+                if (lt->localizedText.locale.length < 2 ||
+                        (lt->localizedText.locale.length > 2 &&
+                         lt->localizedText.locale.data[2] != '-'))
+                    continue;
+
+                UA_String currentPrefix;
+                currentPrefix.data = lt->localizedText.locale.data;
+                currentPrefix.length = 2;
+
+                if (UA_String_equal(&requestedPrefix, &currentPrefix)) {
+                    UA_LocalizedText_copy(&lt->localizedText, &result);
+                    return result;
+                }
+            }
+        }
+    }
+
+    lt = SLIST_FIRST(&head->description);
+    if (lt) {
+        UA_LocalizedText_copy(&lt->localizedText, &result);
+        return result;
+    } else {
+        return result;
+    }
+}
+
+UA_StatusCode UA_NodeStore_insertOrUpdateDescriptionForHead(const UA_LocalizedText *value,
+                                                            UA_NodeHead *head)
+{
+    LocalizedTextEntry *lt = NULL;
+    UA_Boolean found = false;
+    SLIST_FOREACH(lt, &head->description, pointers) {
+        if (UA_String_equal(&lt->localizedText.locale, &value->locale)) {
+            UA_LocalizedText_clear(&lt->localizedText);
+
+            if (value->text.length == 0) {
```

Comment:
I don't understand this.
If `value->text.length == 0` the lookup will never match.
Except for the odd case where an empty description was already stored.

Reply:
This is meant to allow removing a translation from the list by passing an empty text with the locale to remove.